### PR TITLE
[RLlib] Added DefaultCallbacks which replaces old callbacks dict interface

### PR DESCRIPTION
**Pull request by roireshef**
## Why are these changes needed?

Those additions are not mandatory, although useful to remove boilerplate and arrange the metric callbacks code in a more modular manner - decoupling each metric's code into a separate class such that metrics can be added and removed with ease.

## Related issue number

No issue here :) just a nice feature

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

**Review by sven1977 — COMMENTED** [review]
This looks like a super useful thing. Thanks!

File: rllib/metrics/metric_wrapper.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+             config = {
+                ...
+                "callbacks": metrics_wrapper.to_dict()
+             }
+     """
+    def __init__(self, metrics: List[Metric]):
+        self.metrics = metrics
+
+    def to_dict(self):
```
Comment (sven1977) [inline rllib/metrics/metric_wrapper.py]: Probably need to add `on_sample_end`, `on_train_result`, `on_postprocess_traj`.

**Review by sven1977 — COMMENTED** [review]
Could you add docstrings to the `Metric` class methods? Users need to know, when exactly each of the methods gets called (kind of explicatory from the method name, but still) and what the keys of `info` would hold and where the data in there exactly comes from.

**Comment by sven1977** [comment]
Also, the LINT test in the Travis CI suite complained about two small things:

./rllib/examples/custom_metrics_with_metric_wrapper.py:1:80: E501 line too long (89 > 79 characters)

./rllib/metrics/metric_wrapper.py:9:80: E501 line too long (102 > 79 characters)

**Comment by roireshef** [comment]
Jenkins retest this please

**Comment by roireshef** [comment]
@sven1977 took care of all of your comments. Please note that I didn't write the API for those callbacks, just used it. The callbacks aren't well documented anywhere that I know, so my effort trying to provide a documentation for the Metric abstract callbacks is based on reverse-engineering those calls. I think it would be a good idea for one of the admins (or anyone who's been involved with defining the callbacks API) to verify the correctness of my docstrings.

File: rllib/metrics/metric.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
+from abc import abstractmethod
+from typing import Dict
+
+
+class Metric:
```
Comment (ericl) [inline rllib/metrics/metric.py]: IIUC this is more of a CallbacksWrapper (similar to the Keras callbacks class). Since the callbacks can be used for more than recording metrics (e.g., can be used for curriculum learning and other ad hoc model updates), how about we do the following renaming:
- Metric => Callback
- MetricsWrapper => CallbacksList

I think this would make it more clear for users. We can even point to https://keras.io/callbacks/ for conceptual guidance.

File: rllib/metrics/metric.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    @abstractmethod
+    def on_episode_start(info: Dict) -> None:
+        """
+        Runs before an episode starts. This is where buffer initialization and
+        counter resetting is being done.info["episode"].user_data is the
+        natural place to store intermediate data
+        :param info: dictionary containing the following keys:
+            "env": ray.rllib.BaseEnv instance running the episodes
+            "policy": Dict[str, Policy] - mapping all available policies
+            "episode": MultiAgentEpisode - see class
```
Comment (ericl) [inline rllib/metrics/metric.py]: These docstrings are very useful. Should we lift them to top level params?, e.g.
```
def on_episode_start(env: BaseEnv, policy: Dict[str, Policy], episode: MultiAgentEpisode):
    """Runs before an episode starts.

    Args:
        env: ray.rllib.BaseEnv instance running the episodes.
        policy: Dict[str, Policy] - mapping all available policies.
        episode: MultiAgentEpisode - the current episode state.
    """

**Comment by ericl** [comment]
Sorry I took this over a bit with refactoring; now it replaces and deprecates the old callbacks interface.

cc @sven1977 @richardliaw

File: .gitignore [hunk 4]
```diff
@@ -106,6 +106,7 @@ scripts/nodes.txt
 # Generated documentation files
 /doc/_build
 /doc/source/_static/thumbs
+doc/source/tune/generated_guides/
```
Comment (richardliaw) [inline .gitignore:109]: /doc?

File: doc/source/rllib-training.rst [hunk 5]
```diff
@@ -496,51 +496,12 @@ Ray actors provide high levels of performance, so in more complex cases they can
...
-        pole_angle = abs(episode.last_observation_for()[2])
-        episode.user_data["pole_angles"].append(pole_angle)
-
-    def on_episode_end(info):
-        episode = info["episode"]
-        pole_angle = np.mean(episode.user_data["pole_angles"])
-        print("episode {} ended with length {} and pole angles {}".format(
-            episode.episode_id, episode.length, pole_angle))
-        episode.custom_metrics["pole_angle"] = pole_angle
+User-defined state can be stored for the `episode <https://github.com/ray-project/ray/blob/master/rllib/evaluation/episode.py>`__ in the ``episode.user_data`` dict, and custom scalar metrics reported by saving values to the ``episode.custom_metrics`` dict. These custom metrics will be aggregated and reported as part of training results. For a full example, see `custom_metrics_and_callbacks.py <https://github.com/ray-project/ray/blob/master/rllib/examples/custom_metrics_and_callbacks.py>`__.
```
Comment (richardliaw) [inline doc/source/rllib-training.rst:501]: It would be helpful if you showed when these callbacks were invoked in the event loop.

File: rllib/agents/callbacks.py [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+    def on_episode_start(self, worker: RolloutWorker, base_env: BaseEnv,
+                         policies: Dict[str, Policy],
+                         episode: MultiAgentEpisode, **kwargs):
+        """Callback run on the rollout worker before each episode starts.
+
+        Args:
+            worker (RolloutWorker): Reference to the current rollout worker.
+            base_env (BaseEnv): BaseEnv running the episode. The underlying
+                env object can be gotten by calling base_env.get_unwrapped().
+            policies (dict): Mapping of policy id to policy objects.
```
Comment (richardliaw) [inline rllib/agents/callbacks.py:38]: Problem with `policy id` is that I don't actually know how to interact with this... do I use `default`?

File: rllib/agents/trainer.py [hunk 7]
```diff
@@ -578,6 +561,7 @@ def normalize(env):
             self.env_creator = lambda env_config: normalize(inner(env_config))
 
         Trainer._validate_config(self.config)
+        self.callbacks = self.config["callbacks"]()
```
Comment (richardliaw) [inline rllib/agents/trainer.py:564]: consider doing a type check here to provide a better user error

**Comment by richardliaw** [comment]
Callback class is much cleaner... consider adding tests

**Comment by richardliaw** [comment]
Will leave it to @sven1977 to read the code ;)

File: doc/source/rllib-training.rst (hunk 5 above)
Comment (ericl) [reply doc/source/rllib-training.rst:501]: Was trying to get "self.callbacks.on" github search to work, but apparently they don't support exact match yet...

File: rllib/agents/callbacks.py (hunk 6 above)
Comment (ericl) [reply rllib/agents/callbacks.py:38]: Yeah I think it's fairly obvious, if you print it out.

File: rllib/agents/trainer.py (hunk 7 above)
Comment (ericl) [reply rllib/agents/trainer.py:569]: Done

**Review by ericl — COMMENTED** [review]
The example is pretty complete in test coverage, though not perfect. I think it's good enough for now, since it was successful in testing callback coverage up to this point.

File: rllib/agents/callbacks.py (hunk 6 above)
Comment (ericl) [reply rllib/agents/callbacks.py]: Added an tip in the doc comments around this.

**Review by ericl — APPROVED** [review]
